### ISSUE TITLE
Feature request: use mpv profile=opengl-hq presets as default

### ISSUE THREAD

**Issue by bodayw** [lhc70000/iina#587]
**System and IINA version:**

- macOS 10.12.4
- IINA 0.0.7

**Expected behavior:**

As briefly discussed in the group chat...

The default mpv opengl renderer settings are rather conservative, and they have built in a profile which contains presets for high quality rendering, as follows (as in 0.24.0):
```
scale=spline36
cscale=spline36
dscale=mitchell
dither-depth=auto
correct-downscaling=yes
sigmoid-upscaling=yes
deband=yes
```
These settings would provide visually much better results, and should run smoothly on most modern GPUs (something like HD 4000 or newer). People with high-end GPUs can always opt for even better but more demanding scaling algorithms (e.g. scale=ewa_lanczossharp), but IMHO the hq presets is a better balance point for most people.

mpv's documentation also recommend using hq presets, as they stated "By default, it tries to use fast and fail-safe settings. Use the ``opengl-hq`` profile to use this driver with defaults set to high quality rendering."

I am thinking of making the hq preset as default, but leave it as optional in the settings panel to let people uncheck it if they don't want it (for reasons like legacy hardware, or battery life concerns when they're using a laptop on the go). Or if you have the same reasoning of mpv's choice of default, at least the hq preset should be made into the settings panel and let people easily switch to the hq presets, rather than type in commands in the advanced settings.

**Comment by lucid-dreams** [comment]
I second that regarding the scaler.

The scaler spline36 is basically free with almost all cards in current use.
I think IINA should not look worse than Quicktime, out of the box.

Also I'm not sure whether color management is done correctly at the moment. The profile opengl-hq also used to include icc-auto, but that does not seem to have any effect.

**Comment by bodayw** [comment]
Now that I gave it a second thought, I think it's better to leave the default as in mpv, but I do think it's a good idea to add an option in settings to allow people easily apply it. So I've changed the title of the issue.

The main reason I think the default settings still make sense for most people is that Apple has basically made the entire Mac line using "retina" displays, and to watch HD videos in full screen on a very high resolution screen, the hq profile actually put a significantly heavier charge on the system than the default settings. Also, I think in most cases, using bilinear for scaling is just fine, especially when watching HD contents on a laptop screen, the difference is barely noticeable at a normal distance.

I also mentioned color management with @lhc70000 in the group chat, and the speculation was that the released build probably doesn't have littlecms compiled in, which is required for enabling color management in mpv. I haven't had time to try compiling iina on my own, but if this is the case, I would also suggest to have littlecms always included with iina. It's always nice to have color management, and it's not that rare people have their monitors profiled.

**Comment by kkanungo17** [comment]
Wouldn't the option profile=opengl-hq directly work in the advanced mpv options? As for the color management, opengl-cb doesn't support icc-profile-auto, but icc-profile=/path/to/file is supported and works.

**Comment by bodayw** [comment]
>Wouldn't the option profile=opengl-hq directly work in the advanced mpv options?

Yes that works. Just thought it might be good to have it in the settings.

>opengl-cb doesn't support icc-profile-auto, but icc-profile=/path/to/file is supported and works.

Thanks, that's good to know.

**Comment by csdesigns** [comment]
>> Wouldn't the option profile=opengl-hq directly work in the advanced mpv options?

>Yes that works. Just thought it might be good to have it in the settings.

I have not been able to get this to work in the recent builds through the 'Additional MPV options' dialog box. Though I may be inputting the syntax incorrectly: for the **Name** field, I enter _profile_, and for the **Value** field, I enter _opengl-hq_. Upon restart of the IINA app, I receive the following error message: 

`Error setting option --profile=opengl-hq with return value -5. Please check your extra option settings in Preference > Advanced.`

If I manually write _profile=opengl-hq_ to mpv's config file however, then I have no issue and rendering is correct. I have verified opengl-hq is enabled through config file usage by playing one of my 10bit HEVC test streams, as without using opengl-hq the rendering is 8bit-only (or non-dithered), vs the result with this enabled produces the smooth gradients in the playback I am expecting.

**Comment by inflation** [comment]
Very strangely, even though I have `profile=opengl-hq` in my mpv.conf, IINA still uses `opengl-cb` for some reason.

Log here: https://clbin.com/BhOtO

**Comment by kkanungo17** [comment]
Maybe because opengl-* options have now been superseded by gpu-* options. So now the syntax would be --profile=gpu-hq.

**Comment by ElhemEnohpi** [comment]
I can confirm that adding an entry to "Additional mpv options" with `name: profile` and `value: gpu-hq` (not opengl-hq) works as expected. I'm new to IINA, and was quite disappointed with the image quality, particularly very noticeable banding in shadow areas, compared to mpv. Enabling gpu-hq fixed it. I think it should be the default, but if not, then a prominent option in the settings.

**Comment by kkanungo17** [comment]
GPU-HQ is overkill in most cases, especially when current Macs have retina displays, incurring even higher cost. Only deband, dithering and icc profile should have their own options in settings IMO.

**Comment by inflation** [comment]
@ElhemEnohpi The VO IINA (and newer version mpv on macOS) used is `libmpv`, not quite `gpu`, and `gpu-hq` only tweaks several options to make it rendering in higher quality (but not ultimate). If you want deband algorithm, just set `deband` to `yes`. If you want best rendering pictures, you should manually set those options either.

Personally, I use a script to switch profiles between different situations, e.g., battery, resolution, etc. But for most people, fast and fail-safe settings are good enough.

BTW, @bodayw color management is in `develop` now.

**Comment by ElhemEnohpi** [comment]
@inflation Thanks for your comment, though I have to say I didn't understand most of it. All I know is that IINA looked very bad compared to mpv, until I set `gpu-hq`. I do understand that `gpu-hq` only tweaks some options, as noted in the first post. My old 2012 MacBook Air plays back in this configuration using less than 50% cpu even in software-only (eg. HEVC) playback. So maybe it is "overkill" but for me anyway it's not a problem, and works well, with very noticeably improved quality over the default.

I trust that the options set by `gpu-hq` have been chosen by the developers to give a reasonable "high quality" option. I must disagree that "for most people, fast and fail-safe settings are good enough". At least to my eye they are not at all good enough! Probably you are correct that manually setting all the different options could give a better performance. But it becomes too difficult for the average user (including me) to understand all the options. I think a more simple choice in the GUI between "use fast and fail-safe settings" (which includes for example "very low quality" bilinear scaling), and "high quality rendering", makes sense for a Mac application. The ability to override all settings, for people interested in ultimate optimization, is still possible through manual advanced options settings.

**Comment by kkanungo17** [comment]
> @inflation Thanks for your comment, though I have to say I didn't understand most of it. All I know is that IINA looked very bad compared to mpv, until I set `gpu-hq`. I do understand that `gpu-hq` only tweaks some options, as noted above in the first post. My old 2012 MacBook Air plays back in this configuration using less than 50% cpu even in software-only (eg. HEVC) playback. So maybe it is "overkill" but for me anyway it's not a problem, and works well, with very noticeably improved quality over the default.
> 
> I trust that the options set by `gpu-hq` have been chosen by the developers to give a reasonable "high quality" option. I must disagree that "for most people, fast and fail-safe settings are good enough". At least to my eye they are not at all good enough! Probably you are correct that manually setting all the different options could give a better performance. But it becomes too difficult for the average user (including me) to understand all the options. I think a more simple choice in the GUI between "use fast and fail-safe settings" (which includes for example "very low quality" bilinear scaling), and "high quality rendering", makes sense for a Mac application. The ability to override all settings, for people interested in ultimate optimization, is still possible through manual advanced options settings.

IINA default is the same as mpv default though, except maybe the keybindings.

**Comment by ElhemEnohpi** [comment]
> IINA default is the same as mpv default though, except maybe the keybindings.

You're right about that. Sorry, maybe I'm not explaining this very well. I've been using mpv until now, and I did spend a lot of time to read the dozens of pages of documentation, and the hundreds of options. I think that's kind of expected with mpv. Despite knowing a fair amount about computers and video, I still didn't understand a lot of it. But I learned enough to set "gpu-hq". At the moment an IINA user would have to go through all that too, which is maybe too much to ask, if the intention is to be a more Mac-like, mainstream app.

My point is only that the current bare-bones default is in my opinion too low-quality, and while it might be "good enough" for some people, I think many would want something better, even if it costs a little more power. In order to improve it by manually setting options though, the average person would need to spend hours or days studying to learn what they all mean. `gpu-hq` gives a fairly simple shortcut that is, for me anyway, "better enough".

It's not necessary to only choose either the existing default or hq profiles though. Maybe three or four carefully-chosen profiles, suited to Mac hardware, could be created and presented to the user as a kind of "quality vs. power" slider or some other reasonably simple choice, for those who don't want to delve deeply into understanding video processing.

I don't know enough to suggest exactly what the profiles should be, or what to do about retina screens. I can only suggest that there could be some simpler way to choose higher quality video settings, whatever that is.

**Comment by kkanungo17** [comment]
I do get where you're coming from, but I really doubt the improvements you're seeing come from all the options that gpu-hq sets. The most difference comes from setting a icc profile and turning on debanding and dithering. Rest of it has a very slight difference, but does have a noticeable hit on battery life. You could maybe add a gamma multiplier to have the contrast match to QuickTime, but that's about all that's required.

**Comment by ElhemEnohpi** [comment]
Thanks, I'm sure you're right about that. I just never had the time or really the need to go through the individual options and test the differences, as gpu-hq was "better enough", but I may try your advice now. Again, it's nice to have a profile of quality options set by an expert, rather than have to figure everything out myself.

**Comment by Tsubajashi** [comment]
is something still happening about that "issue"?
i would love to use gpu-hq on iina. the output looks different compared to mpv + gpu-hq.
is there any way to use it right now?

**Comment by ElhemEnohpi** [comment]
@Tsubajashi , previously you could set it in "advanced" options, see https://github.com/iina/iina/issues/587#issuecomment-424571190 but that doesn't seem to be working for me anymore. I also tried putting it in a .conf file, which didn't work either. So now the video quality of IINA is significantly worse than mpv for me.

**Comment by bodayw** [comment]
@ElhemEnohpi 

I haven't been using IINA for a while so I don't know if those advanced options still work or not.

>Not sure how much difference the other options make, for example the scaling.

You might want to check out some of the comparison screenshots people have made for this purpose, for example:
https://github.com/mpv-player/mpv/wiki/Upscaling

>and I'm not sure whether I should use dithering on my 8-bit LCD monitor, or if it does it itself.

Here dithering refers to those during the rendering process of mpv (e.g. when there is conversion from higher to lower precision), and is irrelevant to the bit-depth of your monitor. It's always preferred for higher visual quality.

**Comment by ElhemEnohpi** [comment]
@bodayw Thanks for the link about upscaling, I'll take a look. About dithering, according to the mpv documentation, `--dither-depth` defaults to "No", and it says:
> Note that the depth of the connected video display device cannot be detected. Often, LCD panels will do dithering on their own, which conflicts with this option and leads to ugly output.

I could be wrong, but sounds to me like it's entirely relevant to the bit-depth of the monitor, for example displaying 10-bit video on an 8-bit monitor, and not necessarily preferred to have it turned on, if the monitor does it in hardware.
